mypy/semanal.py

```diff
@@ -323,34 +315,36 @@ def prepare_method_signature(self, func: FuncDef) -> None:
         if not func.is_static:
             if not func.arguments:
                 self.fail('Method must have at least one argument', func)
-            elif func.type:
-                sig = cast(FunctionLike, func.type)
+            elif isinstance(func.type, FunctionLike):
                 if func.is_class:
                     leading_type = self.class_type(self.type)
                 else:
                     leading_type = self_type(self.type)
-                func.type = replace_implicit_first_type(sig, leading_type)
+                func.type = replace_implicit_first_type(func.type, leading_type)
 
-    def is_conditional_func(self, previous: Node, new: FuncDef) -> bool:
-        """Does 'new' conditionally redefine 'previous'?
+    def set_original_def(self, previous: Node, new: FuncDef) -> bool:
+        """If 'new' conditionally redefine 'previous', set 'previous' as original
 
         We reject straight redefinitions of functions, as they are usually
         a programming error. For example:
 
         . def f(): ...
         . def f(): ...  # Error: 'f' redefined
         """
-        return isinstance(previous, (FuncDef, Var)) and new.is_conditional
+        if isinstance(previous, (FuncDef, Var)) and new.is_conditional:
+            new.original_def = previous
+            return True
+        else:
+            return False
 
     def update_function_type_variables(self, defn: FuncDef) -> None:
         """Make any type variables in the signature of defn explicit.
 
         Update the signature of defn to contain type variable definitions
         if defn is generic.
         """
-        if defn.type:
-            functype = cast(CallableType, defn.type)
-            typevars = self.infer_type_variables(functype)
+        if isinstance(defn.type, CallableType):
```

Comment:
I think this might be cleaner with an assert instead.  This changes the meaning a bit.